torch/distributed/_tensor/ops/matrix_ops.py

```diff
@@ -374,3 +374,183 @@ def constant_pad_nd_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrateg
             )
         ]
     )
+
+
+@register_op_strategy(aten._scaled_dot_product_efficient_attention.default)
+def scaled_dot_product_efficient_attention_strategy(
+    mesh: DeviceMesh, op_schema: OpSchema
+) -> OpStrategy:
+    # NOTE: currently we only support some simple strategies to support tensor parallelism
+    q_input_strategy = op_schema.args_schema[0]
+    assert isinstance(q_input_strategy, OpStrategy)
+    # assuming q/k/v have the same shape
+    qkv_shape = q_input_strategy.output_shape
+    has_attn_bias = op_schema.args_schema[3] is not None
+    compute_log_sumexp = op_schema.args_schema[4]
+
+    all_mesh_dim_strategies = []
+
+    for mesh_dim in range(mesh.ndim):
+        single_mesh_dim_strategies = []
+
+        # placement list stores placements of [outputs, inputs]
+        # in the spda case, we have 2 valid tensor outputs and 3 or 4 tensor inputs
+        # first we can always accept full replication for both inputs and outputs
+        all_replicate: List[Placement] = [Replicate()] * (5 + has_attn_bias)
+        single_mesh_dim_strategies.append(all_replicate)
+
+        # second we can accept the sharding pattern of tensor parallelism, which
+        # shard on the heads dimension
+        qkv_sharding = Shard(1)
```

Comment:
is there a way to unify the common part of flash attention op strategy  and efficeint attention startegy to a function and call them differently to reuse the code?

Reply:
I thought about if this is possible. It seems for flash attention and efficient attention, the sets of parameters and their orderings are very different -- apparently the interfaces are not aligned when they design the two attention backends. From my perspective, the part of code they share is very minimal.

So IMO it's better to separate them, especially considering there are efforts like ring attention (#122460) which only works on top of one of them.

Reply:
Hmm I was referring to the ring attention case, where it seems like we can use a templated attention function to encapsulate flash + memory attention https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/experimental/attention.py#L187

But anyways I think this could be a follow up improvement later